src/mlpack/methods/ann/layer/lookup_impl.hpp

```diff
@@ -0,0 +1,185 @@
+Lookup<MatType, RegularizerType>::operator=(const Lookup& layer)
+{
+  if (&layer != this)
+  {
+    Layer<MatType>::operator=(layer);
+    vocabSize = layer.vocabSize;
+    embeddingSize = layer.embeddingSize;
+    regularizer = layer.regularizer;
+  }
+
+  return *this;
+}
+
+template<typename MatType, typename RegularizerType>
+Lookup<MatType, RegularizerType>&
+Lookup<MatType, RegularizerType>::operator=(
+    Lookup&& layer)
+{
+  if (&layer != this)
+  {
+    Layer<MatType>::operator=(std::move(layer));
+    vocabSize = std::move(layer.vocabSize);
+    embeddingSize = std::move(layer.embeddingSize);
+    regularizer = std::move(layer.regularizer);
+
+    // Reset parameters of other layer.
+    layer.vocabSize = 0;
+    layer.embeddingSize = 0;
+  }
+
+  return *this;
+}
+
+template<typename MatType, typename RegularizerType>
+void Lookup<MatType, RegularizerType>::SetWeights(const MatType& weightsIn)
+{
+  MakeAlias(weights, weightsIn, vocabSize, embeddingSize);
+}
+
+template<typename MatType, typename RegularizerType>
+void Lookup<MatType, RegularizerType>::Forward(
+    const MatType& input, MatType& output)
+{
+  const size_t seqLength = input.n_rows;
+  const size_t batchSize = input.n_cols;
+
+  output.set_size(seqLength * embeddingSize, batchSize);
+
+  for (size_t i = 0; i < batchSize; ++i)
+  {
+    //! ith column of output is a vectorized form of a matrix of shape
+    //! (seqLength, embeddingSize) selected as a combination of rows from the
+    //! weights. The MultiheadAttention class requires this particular ordering
+    //! of matrix dimensions.
+    output.col(i) = arma::vectorise(weights.rows(
+        arma::conv_to<arma::uvec>::from(input.col(i)) - 1));
+  }
+}
+
+template<typename MatType, typename RegularizerType>
+void Lookup<MatType, RegularizerType>::Backward(
+    const MatType& /* input */,
+    const MatType& /* output */,
+    const MatType& gy,
+    MatType& g)
+{
+  Log::Fatal << "Lookup cannot be used as an intermediate layer." << std::endl;
+}
+
+template<typename MatType, typename RegularizerType>
+void Lookup<MatType, RegularizerType>::Gradient(
+    const MatType& input,
+    const MatType& error,
+    MatType& gradient)
+{
+  const size_t seqLength = input.n_rows;
+  const size_t batchSize = input.n_cols;
+
+  const CubeType errorTemp;
+  MakeAlias(const_cast<CubeType&>(errorTemp), error, seqLength, embeddingSize,
+      batchSize, 0, false);
+
+  gradient.set_size(arma::size(weights));
+  gradient.zeros();
+
+  for (size_t i = 0; i < batchSize; ++i)
+  {
+    gradient.rows(arma::conv_to<arma::uvec>::from(input.col(i)) - 1)
+        += errorTemp.slice(i);
+  }
+}
+
+template<typename MatType, typename RegularizerType>
+void Lookup<MatType, RegularizerType>::ComputeOutputDimensions()
+{
+  this->outputDimensions = std::vector<size_t>(this->inputDimensions.size(),
+    1);
+
+  // The Linear layer flattens its input.
+  this->outputDimensions[0] = this->embeddingSize * this->inputDimensions[0];
```

Comment:
Don't you need to accumulate the total number of dimensions?  This looks like it will simply ignore any input dimensions past the first.  Alternately, we can require that the number of input dimensions is 1 for the embedding layer; I can see an argument for either case.  (Personally I would allow multidimensional inputs---the existing code will work with that anyway.  We just need to set the output dimension size correctly.)